stdlib/Distributed/src/remotecall.jl

```diff
@@ -548,11 +555,40 @@ Further calls to `fetch` on the same reference return the cached value. If the r
 is an exception, throws a [`RemoteException`](@ref) which captures the remote exception and backtrace.
 """
 function fetch(r::Future)
-    r.v !== nothing && return something(r.v)
-    v = call_on_owner(fetch_ref, r)
-    r.v = Some(v)
+    v_cache = @atomic r.v
+    v_cache !== nothing && return something(v_cache)
+
+    if r.where == myid()
+        v_cache, rv = lock(r.lock) do
+            v_cache = @atomic :monotonic r.v
+            v_cache, v_cache === nothing ? lookup_ref(remoteref_id(r)) : nothing
+        end
+
+        if v_cache !== nothing
+            return something(v_cache)
+        else
+            v_local = fetch(rv.c)
+        end
+    else
+        v_local = call_on_owner(fetch_ref, r)
```

Comment:
This could probably be locked, so that `call_on_owner` only gets called once, but I'm not sure if there's any hidden implications of doing that